prow/config/config.go

```diff
@@ -913,27 +913,32 @@ func setPeriodicDecorationDefaults(c *Config, ps *Periodic) {
 // finalizeJobConfig mutates and fixes entries for jobspecs
 func (c *Config) finalizeJobConfig() error {
 	if c.decorationRequested() {
-		if c.Plank.DefaultDecorationConfig == nil {
-			return errors.New("no default decoration config provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.UtilityImages == nil {
-			return errors.New("no default decoration image pull specs provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.GCSConfiguration == nil {
-			return errors.New("no default GCS decoration config provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.GCSCredentialsSecret == "" {
-			return errors.New("no default GCS credentials secret provided for plank")
-		}
+		if len(c.Plank.DefaultDecorationConfigs) == 0 {
```

Comment:
IMHO we should extend `parseProwConfig` where we currently have most of the ProwConfig defaulting to 

* Have a check if both `c.Plank.DefaultDecorationConfig` and `c.Plank.DefaultDecorationConfigs["*"]` is set and if yes error out
* Afterwards, if `c.Plank.DefaultDecorationConfig` is set, copy it into `c.Plank.DefaultDecorationConfigs["*"]`
* Have no other code that accesses `c.Plank.DefaultDecorationConfig` and change it to use `c.Plank.DefaultDecorationConfigs["*"]` instead

I think this would be a great simplification over this approach here.

Also from a pure logical standpoint, the validation right now is incomplete. The assumption so far has been that all decoration requests can be satisfied with one DecorationConfig, which is why the whole `if c.decorationRequested() ` was suffient.

This is not true anymore, we basically have to check for every single decoration request if we can satisfy it.

Reply:
The current changes in this PR, covers almost all of the above except allowing both `old` and `new` configs. I will address that. 

Also, we can agree if we want to validate everything. I only validate the defaults, because I follow the same logic that we already have with the decoration configs in prowjob level. We don't validate that and instead, we are merging with the default config.  But as you said, it makes sense to validate everything, so I will address also that. Thanks!

Reply:
I don't think the PR does what Alvaro proposed though. The PR does  _If we do not have new-style config and we have the old style one, it is validated and then assigned into the new-style `"*"` default._. This makes it more complicated than necessary. 

IMO, it is completely reasonable to only support having a single-style config: you can have the old stanza, you can have new, but not both at the same time. Then, like Alvaro suggests, in `parseProwConfig`, if you encounter old-style stanza, you check if new-style stanza is unset (and error out if it is set) and turn the old-style stanza (without validation) into the new-style `"*"` default internally. Later you only need to validate the new-style defaults, and like Alvaro says, we need to check that *nothing* else touches the old-style struct member.

Reply:
Yes, the old default config is being called in a few places.  I will fix that also in this PR.  I am a little worried about putting them into `parseProwConfig` because I will have to refactor the tests that are not related to this PR.  Let me check that and see if the changes are not big.